### PR TITLE
Implemented travel routes options feature

### LINKED ISSUE #11: As a user, I want a list of possible travel routes so I can pick my preferred route
**Acceptance Criteria**
Given that I am on the directions page,
And that I have selected the Bus routes option,
Then I should see different possible travel routes to choose from (a recommended route, a route with least walking, and a route with fewest transfers).

### PR DESCRIPTION
**Pull request by pobudzey**
User story #11 is complete, please let me know if something needs to be improved.

Here's what the final product looks like:

93770790_664213144362287_2477671160234377216_n

### CODE REVIEW

File: app/src/main/java/com/example/campusguide/DirectionsActivity.kt [hunk 1]
```diff
@@ -109,26 +128,60 @@ class DirectionsActivity : AppCompatActivity() {
...
                     if (checked) {
+                        showMap()
                         removePreviousPath()
-                        currentPath = paths.getValue("driving")
+                        currentPath = mainPaths.getValue("driving")
                         setPathOnMapAsync(currentPath)
                     }
                 R.id.radio_walking ->
                     if (checked) {
+                        showMap()
```
Comment (felixlapierre) [inline app/src/main/java/com/example/campusguide/DirectionsActivity.kt]: Lines 131-134 are duplicates of lines 138-141 except the string passed to getValue, can we extract them into a common method?

File: app/src/main/java/com/example/campusguide/DirectionsActivity.kt [hunk 2]
```diff
@@ -109,26 +128,60 @@ class DirectionsActivity : AppCompatActivity() {
...
-                        currentPath = paths.getValue("walking")
+                        currentPath = mainPaths.getValue("walking")
                         setPathOnMapAsync(currentPath)
                     }
                 R.id.radio_transit ->
                     if (checked) {
-                        removePreviousPath()
-                        currentPath = paths.getValue("transit")
-                        setPathOnMapAsync(currentPath)
+                        hideMap()
```
Comment (felixlapierre) [inline app/src/main/java/com/example/campusguide/DirectionsActivity.kt:138]: This chunk of the method gets pretty deep into indentation making it hard to read, can we extract the contents of `if (checked) {` block into its own method?

File: app/src/main/java/com/example/campusguide/DirectionsActivity.kt [hunk 3]
```diff
@@ -109,26 +128,60 @@ class DirectionsActivity : AppCompatActivity() {
...
             }
         }
     }
 
+    /**
+     * Callback method to be invoked when an item in this AdapterView has
+     * been clicked.
+     */
+    override fun onItemClick(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
+        showMap()
```
Comment (felixlapierre) [inline app/src/main/java/com/example/campusguide/DirectionsActivity.kt]: I like how you extracted the showMap, hideMap, functionality into their own methods :D

File: app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        val steps = results[position].steps
+        var n = 3
+        if (steps.size < 4) {
+            n = steps.size - 1
+        }
+        for (i in 0..n) {
+            val stepIcon = ImageView(activity)
+            val stepTransitLine = TextView(activity)
+            when (steps[i].travelMode) {
+                "DRIVING" -> {
```
Comment (felixlapierre) [inline app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt]: Extract this string boi into a constant

File: app/src/main/java/com/example/campusguide/DirectionsActivity.kt [hunk 5]
```diff
@@ -63,33 +72,43 @@ class DirectionsActivity : AppCompatActivity() {
...
+        // Hash map containing (travelMode, path) pairs for the three main paths
+        mainPaths = mapOf(
+            "driving" to createPath(startName, endName, "driving", null),
+            "walking" to createPath(startName, endName, "walking", null),
+            "transit" to createPath(startName, endName, "transit", null)
+        )
+
+        // Hash map containing (title, path) pairs for the optional transit paths
+        extraPaths = mapOf(
+            "RECOMMENDED ROUTE" to mainPaths.getValue("transit"),
```
Comment (felixlapierre) [inline app/src/main/java/com/example/campusguide/DirectionsActivity.kt]: Keys in hashmaps should definitely be in constants!

File: app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+import android.widget.BaseAdapter
+import android.widget.ImageView
+import android.widget.LinearLayout
+import android.widget.TextView
+import com.example.campusguide.R
+import java.text.SimpleDateFormat
+import java.util.Calendar
+import java.util.Locale
+
+class TransitRouteAdapter constructor(private val activity: Activity) : BaseAdapter() {
```
Comment (felixlapierre) [inline app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt:23]: A Javadoc to explain what the class does would be helpful i think, esp cause it's not always obvious what BaseAdapter is for.

File: app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        val row = inflater.inflate(R.layout.transit_route_item, parent, false)
+
+        // Set the route title
+        row.findViewById<TextView>(R.id.routeTitle).text = results[position].title
+
+        // Set the route travel time
+        val travelTime = "${results[position].duration / 60} min"
+        row.findViewById<TextView>(R.id.routeDuration).text = travelTime
+
+        // Display the first four steps
```
Comment (felixlapierre) [inline app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt:42]: Why the first 4 steps? Cause it looks nice?

**Review by felixlapierre — APPROVED** [review]
Good stuff! Please address my few minor comments before merging :)

File: app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt (hunk 7 above)
Comment (pobudzey) [reply app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt:42]: Just so it fits nicely into one line, there's really no point in displaying more steps
I also, admittedly, looked at how Google Maps handled the case of a route containing a lot of steps

File: app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt (hunk 6 above)
Comment (pobudzey) [reply app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt:23]: Agreed, done

File: app/src/main/java/com/example/campusguide/DirectionsActivity.kt (hunk 5 above)
Comment (pobudzey) [reply app/src/main/java/com/example/campusguide/DirectionsActivity.kt]: Done

File: app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt (hunk 4 above)
Comment (pobudzey) [reply app/src/main/java/com/example/campusguide/directions/TransitRouteAdapter.kt]: Done

File: app/src/main/java/com/example/campusguide/DirectionsActivity.kt (hunk 3 above)
Comment (pobudzey) [reply app/src/main/java/com/example/campusguide/DirectionsActivity.kt]: Thank you!

File: app/src/main/java/com/example/campusguide/DirectionsActivity.kt (hunk 1 above)
Comment (pobudzey) [reply app/src/main/java/com/example/campusguide/DirectionsActivity.kt]: Good point, the code now looks much cleaner and understandable

File: app/src/main/java/com/example/campusguide/DirectionsActivity.kt (hunk 2 above)
Comment (pobudzey) [reply app/src/main/java/com/example/campusguide/DirectionsActivity.kt:138]: I think extracting some of the functionality into its own method (from the comment above) made this part of the code short and understandable enough

**Pull request merged**